### PR TITLE
Events - Add `CBA_fnc_addBISPlayerEventHandler` for adding EHs to just the player (from ACE3)

### PR DESCRIPTION
**Pull request by LinkIsGrim**
**When merged this pull request will:**
- Title, see https://github.com/acemod/ACE3/pull/10056, though this is not intended for public usage/API (it's a secret tool that will help us later).

~~~I'm not sure if the code block below is equivalent, since there's no `ACE_player` here.~~~ It isn't.
```
if (_ignoreVirtual && {(unitIsUAV focusOn) || {getNumber (configOf focusOn >> "isPlayableLogic") == 1}}) exitWith {};
```

### CODE REVIEW

File: addons/events/fnc_addPlayerEngineEvent.sqf [hunk 1]
```diff
@@ -0,0 +1,70 @@
+#include "script_component.hpp"
+/* ----------------------------------------------------------------------------
+Function: CBA_fnc_addPlayerEngineEvent
```
Comment (veteran29) [inline addons/events/fnc_addPlayerEngineEvent.sqf:3]: The function adds an Event **Handler** so the name should reflect that.

We have `addBISEventHandler` so something like `addBISPlayerEventHandler` would be fitting. I don't like how the capitalized letters merge though.

**Review by veteran29 — DISMISSED** [review]

**Comment by PabstMirror** [comment]
I think we can replace ace_player with `call CBA_fnc_currentUnit`
iirc focusOn might be funny when controlling a uav, not sure

File: addons/events/fnc_addPlayerEngineEvent.sqf [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    }, false] call CBA_fnc_addPlayerEventHandler;
+};
+
+
+private _key = format [QGVAR(playerEvents_%1), toLowerANSI _type]; // should only have one per event, this isn't public
+if (_key in GVAR(playerEventsHash)) exitWith { ERROR_1("bad key %1",_this); };
+
+GVAR(playerEventsHash) set [_key, [_type, _code, _ignoreVirtual]];
+
+if (_ignoreVirtual && {(unitIsUAV focusOn) || {getNumber (configOf focusOn >> "isPlayableLogic") == 1}}) exitWith {};
```
Comment (PabstMirror) [inline addons/events/fnc_addPlayerEngineEvent.sqf:66]:
```suggestion
private _player = call CBA_fnc_currentUnit;
if (_ignoreVirtual && {(unitIsUAV _player ) || {getNumber (configOf _player >> "isPlayableLogic") == 1}}) exitWith {};
```

File: addons/events/fnc_addPlayerEngineEvent.sqf [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+// Add event now
+private _newEH = focusOn addEventHandler [_type, _code];
```
Comment (PabstMirror) [inline addons/events/fnc_addPlayerEngineEvent.sqf:69]:
```suggestion
private _newEH = _player addEventHandler [_type, _code];
```

File: addons/events/fnc_addPlayerEngineEvent.sqf [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+focusOn setVariable [_key, _newEH];
```
Comment (PabstMirror) [inline addons/events/fnc_addPlayerEngineEvent.sqf:70]:
```suggestion
_player setVariable [_key, _newEH];
```

File: addons/events/fnc_addBISPlayerEventHandler.sqf [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+};
+
+
+private _key = format [QGVAR(playerEvents_%1), toLowerANSI _type]; // should only have one per event, this isn't public
+if (_key in GVAR(playerEventsHash)) exitWith { ERROR_1("bad key %1",_this); };
+
+GVAR(playerEventsHash) set [_key, [_type, _code, _ignoreVirtual]];
+
+private _player = call CBA_fnc_currentUnit;
+if (_ignoreVirtual && {(unitIsUAV _player ) || {getNumber (configOf _player >> "isPlayableLogic") == 1}}) exitWith {};
```
Comment (commy2) [inline addons/events/fnc_addBISPlayerEventHandler.sqf:67]:
```suggestion
if (_ignoreVirtual && {(unitIsUAV _player) || {getNumber (configOf _player >> "isPlayableLogic") == 1}}) exitWith {};
```

File: addons/events/fnc_addBISPlayerEventHandler.sqf [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            if (_ignoreVirtual && _isVirtual) then { continue };
+
+            private _newEH = _newPlayer addEventHandler [_type, _code];
+            _newPlayer setVariable [_x, _newEH];
+        } forEach GVAR(playerEventsHash);
+    }, false] call CBA_fnc_addPlayerEventHandler;
+};
+
+
+private _key = format [QGVAR(playerEvents_%1), toLowerANSI _type]; // should only have one per event, this isn't public
```
Comment (commy2) [inline addons/events/fnc_addBISPlayerEventHandler.sqf:61]: Prefixing the keys is completely unnecessary, as there can be no name collisions here.

File: addons/events/fnc_addBISPlayerEventHandler.sqf [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    _eventType     - Type of event to add. Can be any engine (not mission) event. <STRING>
+    _eventCode     - Code to run when event is raised. <CODE>
+    _ignoreVirtual - Ignore virtual units (spectators, virtual zeus, UAV RC) [optional] (default: false) <BOOLEAN>
+
+Returns:
+    None
+
+Examples:
+    (begin example)
+        ["example", "FiredNear", {systemChat str _this}] call CBA_fnc_addBISPlayerEventHandler
```
Comment (commy2) [inline addons/events/fnc_addBISPlayerEventHandler.sqf:18]: Seems wrong.

File: addons/events/fnc_addBISPlayerEventHandler.sqf [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+/* ----------------------------------------------------------------------------
+Function: CBA_fnc_addBISPlayerEventHandler
+
+Description:
+    Adds an engine event handler just to the controlled entity.
+
+Parameters:
+    _eventType     - Type of event to add. Can be any engine (not mission) event. <STRING>
+    _eventCode     - Code to run when event is raised. <CODE>
+    _ignoreVirtual - Ignore virtual units (spectators, virtual zeus, UAV RC) [optional] (default: false) <BOOLEAN>
```
Comment (commy2) [inline addons/events/fnc_addBISPlayerEventHandler.sqf:11]: Is including 'virtual entities' what the average user wants? If no, the default value should be `true`.

File: addons/events/fnc_addBISPlayerEventHandler.sqf [hunk 9]
```diff
@@ -0,0 +1,71 @@
+#include "script_component.hpp"
+/* ----------------------------------------------------------------------------
+Function: CBA_fnc_addBISPlayerEventHandler
+
+Description:
+    Adds an engine event handler just to the controlled entity.
+
+Parameters:
+    _eventType     - Type of event to add. Can be any engine (not mission) event. <STRING>
```
Comment (commy2) [inline addons/events/fnc_addBISPlayerEventHandler.sqf:9]: There are more non-Mission event handlers in the game: [Scripted Event Handlers](https://community.bistudio.com/wiki/Arma_3:_Scripted_Event_Handlers) or [User Interface Event Handlers](https://community.bistudio.com/wiki/User_Interface_Event_Handlers) to name some.

This works with anything from `addEventHandler` only.

```suggestion
    _eventType     - Type of event to add. Can be any event supported by addEventHandler. <STRING>
```

File: addons/events/fnc_addBISPlayerEventHandler.sqf [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+            private _newEH = _newPlayer addEventHandler [_type, _code];
+            _newPlayer setVariable [_x, _newEH];
+        } forEach GVAR(playerEventsHash);
+    }, false] call CBA_fnc_addPlayerEventHandler;
+};
+
+
+private _key = format [QGVAR(playerEvents_%1), toLowerANSI _type]; // should only have one per event, this isn't public
+if (_key in GVAR(playerEventsHash)) exitWith { ERROR_1("bad key %1",_this); };
```
Comment (commy2) [inline addons/events/fnc_addBISPlayerEventHandler.sqf:62]: Does this mean that only one event of each type can be added? If so, how does this fit the spirit of CBA?

File: addons/main/script_mod.hpp [hunk 11]
```diff
@@ -10,7 +10,7 @@
 #define VERSION_AR  MAJOR,MINOR,PATCHLVL,BUILD
 
 // MINIMAL required version for the Mod. Components can specify others..
-#define REQUIRED_VERSION 2.14
+#define REQUIRED_VERSION 2.16
```
Comment (commy2) [inline addons/main/script_mod.hpp:13]:
```suggestion
#define REQUIRED_VERSION 2.14
```
Without `focusOn`, this can remain 2.14.

**Comment by commy2** [comment]
Just a few thoughts.

- It stylistically correctly omits a version to remove the event handlers.

- It is not thread-safe (`spawn`-pseudo threads), like most the rest of CBA tbf.

- This seems related to common component [addPlayerAction](https://github.com/CBATeam/CBA_A3/blob/master/addons/common/fnc_addPlayerAction.sqf), more so than events component [addPlayerEventHandler](https://github.com/CBATeam/CBA_A3/blob/master/addons/events/fnc_addPlayerEventHandler.sqf). The names are unfortunate.

I don't personally like `CBA_fnc_addPlayerAction` and always thought of it as legacy code. I would always recommend to simply add user actions (and by extension event handlers) to every unit and then to quit early via guard clause. Is the mental overhead required for using such an obscure framework justified considering one could simply write:
```sqf
if (_unit != player) exitWith {};
```
?

**Comment by PabstMirror** [comment]
There is some performance savings by only adding events you need.
e.g. FiredNear could almost be considered n^2 based on unit count

We can make it safe by wrapping in a `CBA_fnc_directCall;` like https://github.com/CBATeam/CBA_A3/blob/master/addons/events/fnc_addEventHandler.sqf

**Review by PabstMirror — APPROVED** [review]

File: addons/events/fnc_addPlayerEngineEvent.sqf (hunk 1 above)
Comment (PabstMirror) [reply addons/events/fnc_addPlayerEngineEvent.sqf]: anyone have a different name suggestion?
otherwise I think it's fine as is

**Pull request merged**